### ISSUE TITLE
Corrupt frames on MacBook Pro

### ISSUE THREAD

**Issue by llamafilm** [CESNET/UltraGrid#127]
I'm having a problem receiving HEVC compressed video on a 2015 MacBook Pro running Big Sur.  Sometimes it shows corrupt frames for a few seconds and then it recovers.  It seems to happen mostly on scenes with lots of detail and fast motion, which suggests a decoding issue, but the log is also showing high packet loss around 10-15% sometimes.
A 2019 Mac Mini running Catalina on the same network does not have this problem, which suggests the laptop is running out of resources, but I can't tell what it is.  CPU and memory usage is relatively low.

This is the receiver output when the issue happens:
```
[1609905545.991] Frame incomplete - substream 0, buffer 3312374: expected 61500 bytes, got 52890.
[hevc @ 0x7fdecd80ac00] Could not find ref with POC 0
[1609905547.227] SSRC 4f954b7e: 9196/9856 packets received (93.3%), 660 lost, max loss 64.
[1609905548.471] [GL] 114 frames in 5.02541 seconds = 22.6847 FPS
[1609905549.048] Video dec stats (cumulative): 4199 total / 4191 disp / 0 drop / 36 corr / 8 missing. FEC noerr/OK/NOK: 4198/26/36
```

With `verbose=8` I get an additional message: `Unable to decode frame due to missing data (RTP TS=2123755038)`

I tried various FEC RS settings but it doesn't seem to help.

In the receiver stats, the `isCorrupted` column does not count these bad frames.

I have read the macOS performance tuning guide but I don't think buffer sizes should matter since this HEVC stream is only 20Mbps.  I tried these more conservative settings so I don't have to change the boot-args.
```
sudo sysctl -w kern.ipc.maxsockbuf=16777216
sudo sysctl -w net.inet.tcp.recvspace=1048576
sudo sysctl -w net.inet.udp.recvspace=1048576
```

When I kill the receiver, Activity Monitor continues to show the 20Mbps bandwidth being consumed by `kernel_task` and the fans are still running loud.  So maybe the network chip is over-worked even though CPU usage is low?

<img width="1440" alt="corrupt screenshot" src="https://user-images.githubusercontent.com/3991046/103715545-e3c1a080-4f8e-11eb-810e-443c59755ac3.png">

I monitored UDP packets for awhile with a `netstat -ss` loop and I don't see any significant drop.

**Comment by alatteri** [comment]
Well that sounds like either packet loss or that machine is too slow to decode.  But more like slow cpu if it is on the same LAN ( NOT wifi!!!!)

**Comment by llamafilm** [comment]
Indeed, I think something about the machine is too slow.  I occasionally see that warning in the log.  I'm just trying to figure out what the bottleneck is.  CPU usage is relatively low

**Comment by dutchman80** [comment]
+1 on this issue, using either software decode `uv -d gl` or hardware decode `uv -d gl --param force-lavd-decoder=hevc_cuvid`

**Comment by MartinPulec** [comment]
@llamafilm how is the H.265 compressed? If NVENC, would it be possible try to use libx265? I was thinking that using single core to decode couldn't be enough in spikes. Also there are missing parameters in the original post altogether :)

**Comment by llamafilm** [comment]
Yes this was nvenc. For my use case, I was not interested in x265 because I can’t spare the CPU power. What’s missing?

**Comment by MartinPulec** [comment]
> What’s missing?

I meant the UltraGrid command-line arguments. But never mind - the use of nvenc already answers that. Anyway, in that case this is possibly a duplicate of #121. The dropped packets are just a consequence and I think it is not a bug by itself - the data are dropped in order to present a frames in a timely manner while trying to keep the latency.

**Comment by llamafilm** [comment]
That's probably related, since nvenc is single-threaded.  However, I was watching Activity Monitor which shows usage for each core and none of the cores were close to 100%.

**Comment by MartinPulec** [comment]
Sure, there can be also other explanation - another possibility come into my mind. I was thinking that when there is some complex scene, the bitrate may soar at that moment. What about if the network cannot cope with that spike?

This could be quite easily tested if you know a bitrate that is safe to your network. You can then set the traffic shaper to cap bitrate to this value - it is parameter `-l`, eg. `-l 20M` for 20 Mbps.

**Comment by MartinPulec** [comment]
Please see #241 – the decoding performance of NVENC encoded HEVC should be now improved (runs in parallel), namely with commit ec702d5c.

**Comment by MartinPulec** [comment]
closing for inactivity, you can reopen if you have updates on that
